### ISSUE TITLE
[sklearn#33666] Docs are not explicit about effect of set_params

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#383]
**Mirrored from [scikit-learn#33666](https://github.com/scikit-learn/scikit-learn/issues/33666)**

Original title: Docs are not explicit about effect of set_params
Original labels: Documentation, API
Created: 2026-04-01T11:12:13Z

---

### Describe the issue linked to the documentation

There is quite a bit of documentation for developing scikit-learn-compatible estimators:
https://scikit-learn.org/stable/developers/develop.html

They mention for example that `__init__` should not validate parameters:
https://scikit-learn.org/stable/developers/develop.html#instantiation

And they mention that calling `.fit()` should set attributes with underscores as needed:
https://scikit-learn.org/stable/developers/develop.html#estimated-attributes

But they aren't explicit about one important detail: is the object state after fitting meant to be alterable by `.set_params()`? And if so, should that trigger re-calculations of fitted attributes for example?

For example, should the following sequences be equivalent in order for an estimator to be compatible with scikit-learn?
* `estimator.fit(...).set_output(...).predict(...)`
* `estimator.fit(...).predict(...)`

As far as I can see, most scikit-learn estimators that have some behavior that depends on input parameters after fitting make a copy of the attribute that they used with an underscore - e.g. KNN classes save an 'effective metric', PCA saves `n_components_`, and so on.

But there are cases where it doesn't work like that - e.g.:
```python
import numpy as np
rng = np.random.default_rng(seed=123)
X = rng.standard_normal(size=(1000,10))

from sklearn.neighbors import LocalOutlierFactor

LocalOutlierFactor().fit(X).set_params(novelty=True).predict(X)
# vs.
LocalOutlierFactor().fit(X).predict(X)
```

### Suggest a potential alternative/fix

Would be quite helpful if the doc could be more explicit about how `set_params` should or shouldn't affect the estimator state after a first call to `.fit()` has been made.

Imagine for example a case where calls to `.predict()` use different attributes according to some input parameter that's settable through `.set_params()`.

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Addressed in upstream PR(s). See scikit-learn/scikit-learn PRs for #33666.
